modules/imgproc/src/imgwarp.cpp

```diff
@@ -1830,42 +1831,89 @@ void cv::convertMaps( InputArray _map1, InputArray _map2,
+    CV_Assert( dstm1type == CV_16SC2 || dstm1type == CV_32FC1 || dstm1type == CV_32FC2 );
 
-    if( m2type == CV_16SC2 )
+    if (dstm1type == CV_32FC1)
+    {
+        outputRemapType = RemapType::fp32_mapx_mapy;
+    }
+    else if (dstm1type == CV_32FC2)
+    {
+        outputRemapType = RemapType::fp32_mapxy;
+    }
+    else if (dstm1type == CV_16SC2 && nninterpolate)
+    {
+        outputRemapType = RemapType::int16;
+    }
+    else if (dstm1type == CV_16SC2 && !nninterpolate)
+    {
+        outputRemapType = RemapType::fixedPointInt16;
+    }
+    else if (dstm1type == CV_32SC2 && nninterpolate)
     {
-        std::swap( m1, m2 );
-        std::swap( m1type, m2type );
+        outputRemapType = RemapType::int32;
     }
+    else if (dstm1type == CV_32SC2 && !nninterpolate)
+    {
+        outputRemapType = RemapType::fixedPointInt32;
+    }
+    if (outputRemapType == RemapType::errorType)
+        CV_Error(cv::Error::StsBadSize, errorRemapMessage);
 
-    if( dstm1type <= 0 )
-        dstm1type = m1type == CV_16SC2 ? CV_32FC2 : CV_16SC2;
-    CV_Assert( dstm1type == CV_16SC2 || dstm1type == CV_32FC1 || dstm1type == CV_32FC2 );
     _dstmap1.create( size, dstm1type );
     dstmap1 = _dstmap1.getMat();
 
-    if( !nninterpolate && dstm1type != CV_32FC2 )
+    if (outputRemapType == RemapType::fp32_mapx_mapy || outputRemapType == RemapType::fixedPointInt16 ||
+        outputRemapType == RemapType::fixedPointInt32)
     {
-        _dstmap2.create( size, dstm1type == CV_16SC2 ? CV_16UC1 : CV_32FC1 );
+        _dstmap2.create( size, outputRemapType == RemapType::fp32_mapx_mapy ? CV_32FC1 : CV_16UC1);
         dstmap2 = _dstmap2.getMat();
     }
     else
         _dstmap2.release();
 
-    if( m1type == dstm1type || (nninterpolate &&
-        ((m1type == CV_16SC2 && dstm1type == CV_32FC2) ||
-        (m1type == CV_32FC2 && dstm1type == CV_16SC2))) )
+    if (inputRemapType == outputRemapType)
+    {
+        m1->copyTo(dstmap1);
+        if (!dstmap2.empty())
+            m2->copyTo(dstmap2);
+        return;
+    }
+
+    if (inputRemapType == RemapType::fixedPointInt16 && outputRemapType == RemapType::int16)
+    {
+        m1->copyTo(dstmap1);  // TODO: improved accuracy for inputRemapType fixedPointInt16
+        return;
+    }
+
+    if (inputRemapType == RemapType::fixedPointInt32 && (outputRemapType == RemapType::int16 ||
+                                                         outputRemapType == RemapType::fp32_mapxy))
+    {
+        m1->convertTo(dstmap1, dstmap1.type());  // TODO: improved accuracy
+        return;
+    }
+
+    if ((inputRemapType == RemapType::int16 || inputRemapType == RemapType::int32) &&
+        (outputRemapType == RemapType::fixedPointInt16 || outputRemapType == RemapType::int16 ||
+         outputRemapType == RemapType::fp32_mapxy))
+    {
+        m1->convertTo(dstmap1, dstmap1.type());
+        if (outputRemapType == RemapType::fixedPointInt16)
+            dstmap2 = Mat::zeros( size, CV_16UC1);
```

Comment:
[If I understand correctly](https://docs.opencv.org/4.x/d3/d63/classcv_1_1Mat.html#a56daa006391a670e9cb0cd08e3168c99), a new matrix is allocated only if `dstmap2` has another size or type. In this case `dstmap2` must be zeros and has necessary size (user can use non zeros `dstmap2` or use wrong size of `dstmap2`).

should the assertions be added to avoid this?